### PR TITLE
[AIRFLOW-6850] Handle no access to code in UI

### PR DESCRIPTION
**Pull request by anitakar**
---
Issue link: [AIRFLOW-6850](https://issues.apache.org/jira/browse/AIRFLOW-6850)

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

File: tests/www/test_views.py [hunk 1]
```diff
@@ -569,6 +569,11 @@ def test_code(self):
         resp = self.client.get(url, follow_redirects=True)
         self.check_content_in_response('example_bash_operator', resp)
 
+    def test_code_no_file(self):
+        url = 'code?dag_id=example_bash_operator_xxx'
+        resp = self.client.get(url, follow_redirects=True)
```
Comment (ashb) [inline tests/www/test_views.py]: This error is dag not found (i.e. it's not even in the database) which is very different to dag _source_ code could not be read by webserver. i.e. this test isn't checking what you describe in the PR.

(This request should result in a 302 back to home and show a flash of "DAG example_bash_operator_xxx not found"  or similar)

File: airflow/www/views.py [hunk 2]
```diff
@@ -528,8 +528,9 @@ def code(self, session=None):
                 code = f.read()
             html_code = highlight(
                 code, lexers.PythonLexer(), HtmlFormatter(linenos=True))
-        except OSError as e:
-            html_code = str(e)
+        except Exception as e:
```
Comment (ashb) [inline airflow/www/views.py]: This is a very broad exception. Why the change?

File: airflow/www/views.py [hunk 3]
```diff
@@ -528,8 +528,9 @@ def code(self, session=None):
...
+            html_code = 'Failed to load file ' + (dag.fileloc if dag else 'unknown') +\
+                        '.<br/> Details: ' + str(e)
```
Comment (ashb) [inline airflow/www/views.py]: This is a possible HTML escaping vulnerability here. Do not pass untrusted user input of any form to the UI without HTML escaping it.

File: airflow/www/templates/airflow/dag_code.html [hunk 4]
```diff
@@ -31,7 +31,7 @@ <h4>{{ title }}</h4>
 {% endblock %}
 
 {% block tail %}
-    {{ super() }}
+    {{ super() if dag else '' }}
```
Comment (ashb) [inline airflow/www/templates/airflow/dag_code.html]: I don't see why this change is needed.

**Review by ashb — CHANGES_REQUESTED** [review]
You are conflating dag not found with dag could not be read by webserver.

File: tests/www/test_views.py (hunk 1 above)
Comment (anitakar) [reply tests/www/test_views.py]: Yes, sorry, I was being lazy. I did not know how to write the test I needed so I wrote this test instead. I will try to figure it out.

File: airflow/www/views.py (hunk 2 above)
Comment (anitakar) [reply airflow/www/views.py]: Because with the test I got AttributeError None object doe not have dag_id.
This was me trying to patch it up.

File: airflow/www/templates/airflow/dag_code.html (hunk 4 above)
Comment (anitakar) [reply airflow/www/templates/airflow/dag_code.html]: Otherwise dag page rendering fails on accessing dag attributes when dag is None.

File: airflow/www/views.py (hunk 3 above)
Comment (anitakar) [reply airflow/www/views.py]: True. I did not think about it. Thanks for spotting!

File: airflow/www/views.py [hunk 5]
```diff
@@ -528,8 +528,8 @@ def code(self, session=None):
                 code = f.read()
             html_code = highlight(
                 code, lexers.PythonLexer(), HtmlFormatter(linenos=True))
-        except OSError as e:
-            html_code = str(e)
+        except (OSError, AttributeError) as e:
```
Comment (anitakar) [inline airflow/www/views.py]: I had to add this for test to pass.
io.open(f, mode=mode) line in www/utils.py l 222 fails with AttributeError
    with wwwutils.open_maybe_zipped(dag.fileloc, 'r') as f:
AttributeError: __enter__

File: airflow/www/views.py [hunk 6]
```diff
@@ -529,7 +529,7 @@ def code(self, session=None):
             html_code = highlight(
                 code, lexers.PythonLexer(), HtmlFormatter(linenos=True))
         except OSError as e:
-            html_code = str(e)
+            html_code = 'Failed to load file.<br/> Details: ' + str(e)
```
Comment (ashb) [inline airflow/www/views.py]:
```suggestion
            html_code = 'Failed to load file.<br/> Details: ' + escape(str(e))
```

Should use https://flask.palletsprojects.com/en/1.1.x/api/?highlight=safe#flask.escape

**Review by ashb — APPROVED** [review]

**Review by kaxil — APPROVED** [review]

**Comment by kaxil** [comment]
thanks @anitakar
